### PR TITLE
Install only headers required by julia.h

### LINKED ISSUE #17657: Installs unneeded headers to <prefix>/include/julia?
In the process of packaging julia/0.5.0~rc0 for Debian, I noticed that the runtime library and headers are now installed system-wide for use by other programs. This is a list of headers currently installed here:

```
/usr/include/julia/END.h
/usr/include/julia/ENTRY.amd64.h
/usr/include/julia/ENTRY.i387.h
/usr/include/julia/MurmurHash3.h
/usr/include/julia/arraylist.h
/usr/include/julia/bitvector.h
/usr/include/julia/dirpath.h
/usr/include/julia/dtypes.h
/usr/include/julia/hashing.h
/usr/include/julia/htable.h
/usr/include/julia/ios.h
/usr/include/julia/julia.h
/usr/include/julia/julia_threads.h
/usr/include/julia/julia_version.h
/usr/include/julia/libsupport.h
/usr/include/julia/platform.h
/usr/include/julia/ptrhash.h
/usr/include/julia/strtod.h
/usr/include/julia/timefuncs.h
/usr/include/julia/tzfile.h
/usr/include/julia/utf8.h
/usr/include/julia/utils.h
/usr/include/julia/uv-errno.h
/usr/include/julia/uv-linux.h
/usr/include/julia/uv-threadpool.h
/usr/include/julia/uv-unix.h
/usr/include/julia/uv-version.h
/usr/include/julia/uv.h
```

Did you verify that all of these headers are in fact needed to link to libjulia.so?

### PR DESCRIPTION
**Pull request by petercolberg**
Closes #17657

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -324,6 +324,31 @@ define stringreplace
...
+	src/support/strtod.h \
+	src/support/timefuncs.h \
+	src/support/utf8.h \
+	src/support/utils.h \
+	src/support/uv-errno.h \
+	src/support/uv-linux.h \
+	src/support/uv-threadpool.h \
+	src/support/uv-unix.h \
+	src/support/uv-version.h \
+	src/support/uv.h
```
Comment (tkelman) [inline Makefile]: these `uv*.h` aren't in `src/support`, they're done on a different line

**Comment by tkelman** [comment]
I think this would have less chance of going stale if we filtered out the specific ones we know we don't need. Or move them do a different location like `src/support/include`

**Comment by petercolberg** [comment]
I was thinking of actually testing the installed libjulia by building the embedding example:
```
cc -I/usr/include/julia -o embedding embedding.c -ldl -lrt -lpthread -lopenlibm -ljulia      
/tmp/cc660rPO.o: In function `main':
embedding.c:(.text+0x1c6): undefined reference to `jl_tls_states'
embedding.c:(.text+0x1e9): undefined reference to `jl_tls_states'
embedding.c:(.text+0x2fe): undefined reference to `jl_tls_states'
embedding.c:(.text+0x309): undefined reference to `jl_tls_states'
embedding.c:(.text+0x31a): undefined reference to `jl_tls_states'
collect2: error: ld returned 1 exit status
```
Any idea why the linker cannot find `jl_tls_states`?

**Comment by tkelman** [comment]
yeah https://github.com/JuliaLang/julia/issues/8757 has been open for a while. I think you need a `-DJULIA_ENABLE_THREADING`

**Comment by petercolberg** [comment]
Now the example compiles but fails to run:
```
./embedding 
ERROR: system image file "/home/peter/git/github.com/JuliaLang/julia/examples/../lib/x86_64-linux-gnu/julia/v0.5/sys.so" not found
```
Which variable sets the path to the system image?

**Comment by tkelman** [comment]
`JULIA_HOME`, are you setting that as an environment variable?

**Comment by petercolberg** [comment]
It seems to work now:
```
JULIA_HOME=/usr ./embedding 
1.4142135623730951
sqrt(2.0) in C: 1.414214e+00
sqrt(2.0) in C: 1.414214e+00
x = [9.000000e+00 8.000000e+00 7.000000e+00 6.000000e+00 5.000000e+00 4.000000e+00 3.000000e+00 2.000000e+00 1.000000e+00 0.000000e+00 ]
my_func(5.0) = 10.000000
UndefVarError(:this_function_does_not_exist)
```
For the purpose of testing this is fine, but why is `JULIA_HOME` not set in `libjulia.so`?

**Comment by tkelman** [comment]
See https://github.com/JuliaLang/julia/blob/643bac816a8b034aed780a507dfed4bf3f87698e/src/init.c#L498, there's a `getenv` to allow you to point libjulia to use a different system image path if you don't have the command-line executable available - so exactly in the embedding case. A lot of people do seem to mistakenly set that variable to the wrong thing and Julia fails to start because of it, so we should document that better or pick a more specific name.

**Comment by petercolberg** [comment]
Have you thought about an `include` dir for public headers, separate from `src` and `src/support`?

**Comment by tkelman** [comment]
The install tree should look like that, but the source tree doesn't necessarily have to.

**Comment by petercolberg** [comment]
Indeed. It is suggestion to make the developer aware that these are public headers.

**Comment by yuyichao** [comment]
Do note that `headers required by julia.h` != `public headers` / `public APIs`. There are functions defined in those headers that are intentionally hidden yet the types defined in the same header are needed to define the ABI.

File: Makefile [hunk 2]
```diff
@@ -324,6 +324,25 @@ define stringreplace
 	$(build_depsbindir)/stringreplace $$(strings -t x - $1 | grep '$2' | awk '{print $$1;}') '$3' 255 "$(call cygpath_w,$1)"
 endef
 
+JL_HEADERS := \
```
Comment (tkelman) [inline Makefile]: My prior feedback is now collapsed, but I still don't like this form where `JL_HEADERS` needs to be kept up to date. A shorter exclusions list of things that are currently included in `src/support/*.h` but aren't needed would be more maintainable.

**Pull request closed without merging**